### PR TITLE
chore: initial protocol api work

### PR DESCRIPTION
**Pull request by ryanthemanuel**
### Additional details

This is an initial base implementation of hooking into the protocol API that will be delivered from the cloud. It hooks into the lifecycle of specs, tests, and browser connections.

### Steps to test

- Clone the `cypress-services` repo
  - Run `yarn`
  - Run `yarn watch` in `packages/app-capture-protocol`
- Set `CYPRESS_LOCAL_PROTOCOL_PATH` to the path to `cypress-services/packages/app-capture-protocol/dist/index.js`
- Clone the `cypress` repo
  - Run `yarn`
  - Execute `cypress run` on a project in record mode with `DEBUG` set to `cypress:*protocol*` and ensure you can see the lifecycle events

### How has the user experience changed?

n/a

### PR Tasks

- [na] Have tests been added/updated?
- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

File: packages/server/lib/cloud/api.ts [hunk 1]
```diff
@@ -241,6 +241,7 @@ export type CreateRunOptions = {
   tags: string[]
   testingType: 'e2e' | 'component'
   timeout?: number
+  protocolManager?: any
```
Comment (mschile) [inline packages/server/lib/cloud/api.ts]: Are we eventually going to type this?

Comment (ryanthemanuel) [reply packages/server/lib/cloud/api.ts]: Oops missed that one.

File: packages/server/lib/cloud/api.ts [hunk 2]
```diff
@@ -330,6 +331,12 @@ module.exports = {
         })
       })
     })
+    .then(async (result) => {
+      // TODO: Get url for the protocol code and pass it down to download
```
Comment (mschile) [inline packages/server/lib/cloud/api.ts]: Suggest namespacing all the TODOs so we can easily verify that all the TODOs have been completed.

```suggestion
      // TODO(protocol): Get url for the protocol code and pass it down to download
```

File: packages/server/lib/cloud/api.ts (hunk 1 above)
Comment (ryanthemanuel) [reply packages/server/lib/cloud/api.ts]: Fixed: [`4f73de7` (#26080)](https://github.com/cypress-io/cypress/pull/26080/commits/4f73de7d7c4162e9cd43dbff03b3fdb1e860f10f)

File: packages/server/lib/cloud/api.ts (hunk 2 above)
Comment (ryanthemanuel) [reply packages/server/lib/cloud/api.ts]: good call.

File: packages/server/lib/cloud/protocol.ts [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+import Debug from 'debug'
+import CDP from 'chrome-remote-interface'
+import type { ProtocolManager } from '@packages/types'
+
+interface AppCaptureProtocolInterface {
+  setupProtocol (url?: string): Promise<void>
+  connectToBrowser (options: any): void
+  beforeSpec (spec: any): void
+  afterSpec (): void
+  beforeTest (test: any): void
```
Comment (mschile) [inline packages/server/lib/cloud/protocol.ts]: Do we want to type any of the `any`s?

File: packages/server/lib/cloud/protocol.ts [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+class ProtocolManagerImpl implements ProtocolManager {
+  private protocol: AppCaptureProtocolInterface | undefined
+
+  protocolEnabled (): boolean {
+    return !!this.protocol
+  }
+
+  async setupProtocol (url?: string) {
+    debug('setting up protocol')
```
Comment (mschile) [inline packages/server/lib/cloud/protocol.ts]: Would it be helpful to include the `url` in the debug log?

File: packages/server/lib/modes/index.ts [hunk 5]
```diff
@@ -5,10 +5,6 @@ import { makeDataContext } from '../makeDataContext'
 import random from '../util/random'
 
 export = (mode, options) => {
-  if (mode === 'record') {
```
Comment (mschile) [inline packages/server/lib/modes/index.ts:8]: Why was this removed?

File: packages/server/lib/modes/run.ts [hunk 6]
```diff
@@ -55,8 +56,8 @@ const relativeSpecPattern = (projectRoot, pattern) => {
   return pattern.map((x) => x.replace(`${projectRoot}/`, ''))
 }
 
-const iterateThroughSpecs = function (options: { specs: SpecFile[], runEachSpec: RunEachSpec, beforeSpecRun?: BeforeSpecRun, afterSpecRun?: AfterSpecRun, config: Cfg }) {
-  const { specs, runEachSpec, beforeSpecRun, afterSpecRun, config } = options
+const iterateThroughSpecs = function (options: { specs: SpecFile[], runEachSpec: RunEachSpec, beforeSpecRun?: BeforeSpecRun, afterSpecRun?: AfterSpecRun, config: Cfg, protocolManager?: any }) {
```
Comment (mschile) [inline packages/server/lib/modes/run.ts]: Add type to `protocolManager`

File: packages/server/lib/modes/run.ts [hunk 7]
```diff
@@ -345,12 +358,13 @@ async function postProcessRecording (options: { quiet: boolean, videoCompression
   return continueProcessing(onProgress)
 }
 
-function launchBrowser (options: { browser: Browser, spec: SpecWithRelativeRoot, setScreenshotMetadata: SetScreenshotMetadata, screenshots: ScreenshotMetadata[], projectRoot: string, shouldLaunchNewTab: boolean, onError: (err: Error) => void, videoRecording?: VideoRecording }) {
-  const { browser, spec, setScreenshotMetadata, screenshots, projectRoot, shouldLaunchNewTab, onError } = options
+function launchBrowser (options: { browser: Browser, spec: SpecWithRelativeRoot, setScreenshotMetadata: SetScreenshotMetadata, screenshots: ScreenshotMetadata[], projectRoot: string, shouldLaunchNewTab: boolean, onError: (err: Error) => void, videoRecording?: VideoRecording, protocolManager?: any }) {
```
Comment (mschile) [inline packages/server/lib/modes/run.ts:361]: Add type to `protocolManager`

File: packages/server/lib/modes/run.ts [hunk 8]
```diff
@@ -443,7 +457,7 @@ function listenForProjectEnd (project, exit): Bluebird<any> {
   })
 }
 
-async function waitForBrowserToConnect (options: { project: Project, socketId: string, onError: (err: Error) => void, spec: SpecWithRelativeRoot, isFirstSpec: boolean, testingType: string, experimentalSingleTabRunMode: boolean, browser: Browser, screenshots: ScreenshotMetadata[], projectRoot: string, shouldLaunchNewTab: boolean, webSecurity: boolean, videoRecording?: VideoRecording }) {
+async function waitForBrowserToConnect (options: { project: Project, socketId: string, onError: (err: Error) => void, spec: SpecWithRelativeRoot, isFirstSpec: boolean, testingType: string, experimentalSingleTabRunMode: boolean, browser: Browser, screenshots: ScreenshotMetadata[], projectRoot: string, shouldLaunchNewTab: boolean, webSecurity: boolean, videoRecording?: VideoRecording, protocolManager?: any }) {
```
Comment (mschile) [inline packages/server/lib/modes/run.ts]:
```suggestion
async function waitForBrowserToConnect (options: { project: Project, socketId: string, onError: (err: Error) => void, spec: SpecWithRelativeRoot, isFirstSpec: boolean, testingType: string, experimentalSingleTabRunMode: boolean, browser: Browser, screenshots: ScreenshotMetadata[], projectRoot: string, shouldLaunchNewTab: boolean, webSecurity: boolean, videoRecording?: VideoRecording, protocolManager?: ProtocolManager }) {
```

File: packages/server/lib/modes/index.ts (hunk 5 above)
Comment (ryanthemanuel) [reply packages/server/lib/modes/index.ts:8]: It's actually dead code. There's a boolean in options that signifies record and we flex the run logic based on it.

File: packages/server/lib/modes/run.ts (hunk 7 above)
Comment (mschile) [inline packages/server/lib/modes/run.ts]:
```suggestion
function launchBrowser (options: { browser: Browser, spec: SpecWithRelativeRoot, setScreenshotMetadata: SetScreenshotMetadata, screenshots: ScreenshotMetadata[], projectRoot: string, shouldLaunchNewTab: boolean, onError: (err: Error) => void, videoRecording?: VideoRecording, protocolManager?: ProtocolManager }) {
```

File: packages/server/lib/modes/run.ts [hunk 9]
```diff
@@ -692,10 +706,10 @@ function screenshotMetadata (data, resp) {
...
+async function runSpecs (options: { config: Cfg, browser: Browser, sys: any, headed: boolean, outputPath: string, specs: SpecWithRelativeRoot[], specPattern: string | RegExp | string[], beforeSpecRun?: BeforeSpecRun, afterSpecRun?: AfterSpecRun, runUrl?: string, parallel?: boolean, group?: string, tag?: string, autoCancelAfterFailures?: number | false, testingType: TestingType, quiet: boolean, project: Project, onError: (err: Error) => void, exit: boolean, socketId: string, webSecurity: boolean, projectRoot: string, protocolManager?: any } & Pick<Cfg, 'video' | 'videoCompression' | 'videosFolder' | 'videoUploadOnPasses'>) {
```
Comment (mschile) [inline packages/server/lib/modes/run.ts]:
```suggestion
async function runSpecs (options: { config: Cfg, browser: Browser, sys: any, headed: boolean, outputPath: string, specs: SpecWithRelativeRoot[], specPattern: string | RegExp | string[], beforeSpecRun?: BeforeSpecRun, afterSpecRun?: AfterSpecRun, runUrl?: string, parallel?: boolean, group?: string, tag?: string, autoCancelAfterFailures?: number | false, testingType: TestingType, quiet: boolean, project: Project, onError: (err: Error) => void, exit: boolean, socketId: string, webSecurity: boolean, projectRoot: string, protocolManager?: ProtocolManager } & Pick<Cfg, 'video' | 'videoCompression' | 'videosFolder' | 'videoUploadOnPasses'>) {
```

File: packages/types/src/server.ts [hunk 10]
```diff
@@ -10,6 +11,7 @@ export type OpenProjectLaunchOpts = {
   videoApi?: RunModeVideoApi
   onWarning: (err: Error) => void
   onError: (err: Error) => void
+  protocolManager?: any
```
Comment (mschile) [inline packages/types/src/server.ts]:
```suggestion
  protocolManager?: ProtocolManager
```

File: packages/types/src/server.ts [hunk 11]
```diff
@@ -21,6 +23,7 @@ export type BrowserLaunchOpts = {
   onBrowserClose?: (...args: unknown[]) => void
   onBrowserOpen?: (...args: unknown[]) => void
   relaunchBrowser?: () => Promise<any>
+  protocolManager?: any
```
Comment (mschile) [inline packages/types/src/server.ts]:
```suggestion
  protocolManager?: ProtocolManager
```

Comment (ryanthemanuel) [reply packages/types/src/server.ts]: [`fcb6c6f` (#26080)](https://github.com/cypress-io/cypress/pull/26080/commits/fcb6c6f9833d27a6958eb0a8dbde8449fabc6447#diff-fae7886d3dd070e07a8b39ac5718534511764b7f03bc24c9726f82977b53a225)

File: packages/types/src/server.ts (hunk 10 above)
Comment (ryanthemanuel) [reply packages/types/src/server.ts]: [`fcb6c6f` (#26080)](https://github.com/cypress-io/cypress/pull/26080/commits/fcb6c6f9833d27a6958eb0a8dbde8449fabc6447#diff-fae7886d3dd070e07a8b39ac5718534511764b7f03bc24c9726f82977b53a225)

File: packages/server/lib/modes/run.ts (hunk 9 above)
Comment (ryanthemanuel) [reply packages/server/lib/modes/run.ts]: [`fcb6c6f` (#26080)](https://github.com/cypress-io/cypress/pull/26080/commits/fcb6c6f9833d27a6958eb0a8dbde8449fabc6447#diff-fae7886d3dd070e07a8b39ac5718534511764b7f03bc24c9726f82977b53a225)

File: packages/server/lib/modes/run.ts (hunk 7 above)
Comment (ryanthemanuel) [reply packages/server/lib/modes/run.ts]: [`fcb6c6f` (#26080)](https://github.com/cypress-io/cypress/pull/26080/commits/fcb6c6f9833d27a6958eb0a8dbde8449fabc6447#diff-fae7886d3dd070e07a8b39ac5718534511764b7f03bc24c9726f82977b53a225)

File: packages/server/lib/modes/run.ts (hunk 8 above)
Comment (ryanthemanuel) [reply packages/server/lib/modes/run.ts]: [`fcb6c6f` (#26080)](https://github.com/cypress-io/cypress/pull/26080/commits/fcb6c6f9833d27a6958eb0a8dbde8449fabc6447#diff-fae7886d3dd070e07a8b39ac5718534511764b7f03bc24c9726f82977b53a225)

File: packages/server/lib/modes/run.ts (hunk 6 above)
Comment (ryanthemanuel) [reply packages/server/lib/modes/run.ts]: [`fcb6c6f` (#26080)](https://github.com/cypress-io/cypress/pull/26080/commits/fcb6c6f9833d27a6958eb0a8dbde8449fabc6447#diff-fae7886d3dd070e07a8b39ac5718534511764b7f03bc24c9726f82977b53a225)

File: packages/types/src/server.ts [hunk 12]
```diff
@@ -90,6 +93,9 @@ export interface OpenProjectLaunchOptions {
   onChange?: WebSocketOptionsCallback
   onError?: (err: Error) => void
 
+  // Manager used to communicate Cypress lifecycle events to the protocol
```
Comment (mschile) [inline packages/types/src/server.ts]: Suggest making it a little more generic:
```suggestion
  // Manager used to communicate with the Cloud protocol
```

File: packages/server/lib/cloud/protocol.ts (hunk 4 above)
Comment (ryanthemanuel) [reply packages/server/lib/cloud/protocol.ts]: [`fcb6c6f` (#26080)](https://github.com/cypress-io/cypress/pull/26080/commits/fcb6c6f9833d27a6958eb0a8dbde8449fabc6447#diff-fae7886d3dd070e07a8b39ac5718534511764b7f03bc24c9726f82977b53a225)

File: packages/server/lib/cloud/protocol.ts (hunk 3 above)
Comment (ryanthemanuel) [reply packages/server/lib/cloud/protocol.ts]: [`fcb6c6f` (#26080)](https://github.com/cypress-io/cypress/pull/26080/commits/fcb6c6f9833d27a6958eb0a8dbde8449fabc6447#diff-fae7886d3dd070e07a8b39ac5718534511764b7f03bc24c9726f82977b53a225)

File: guides/protocol-development.md [hunk 13]
```diff
@@ -0,0 +1,11 @@
...
+
+Under normal circumstances, we retrieve from the cloud the code that is used to capture and communicate test data. In order to develop the code locally, developers will:
+
+* Clone the `cypress-services` repo
+  * Run `yarn`
+  * Run `yarn watch` in `packages/app-capture-protocol`
+* Set `CYPRESS_LOCAL_PROTOCOL_PATH` to the path to `cypress-services/packages/app-capture-protocol/dist/index.js`
+* Clone the `cypress` repo
+  * Run `yarn`
+  * Execute `cypress run` on a project in record mode
```
Comment (mschile) [inline guides/protocol-development.md]:
```suggestion
* Clone the `cypress` repo
  * Run `yarn`
  * Execute `CYPRESS_LOCAL_PROTOCOL_PATH=path/to/cypress-services/packages/app-capture-protocol/dist/index.js CYPRESS_INTERNAL_ENV=staging yarn cypress:run --record --key <record_key> --project <path/to/project>` on a project in record mode
```

File: guides/protocol-development.md [hunk 14]
```diff
@@ -0,0 +1,11 @@
+# Protocol Development
+
+Under normal circumstances, we retrieve from the cloud the code that is used to capture and communicate test data. In order to develop the code locally, developers will:
```
Comment (mschile) [inline guides/protocol-development.md]:
```suggestion
In production, the capture code used to capture and communicate test data will be retrieved from the Cloud. However, in order to develop the capture code locally, developers will:
```

**Review by mschile — APPROVED** [review]
I left a few additional comments but otherwise it looks good!

**Review by mjhenkes — APPROVED** [review]
